GCP CLI support: `tsh` and `tctl` changes

Tener opened the pull request:
`tsh` and `tctl` changes for https://github.com/gravitational/teleport/issues/17257

To be applied on top of https://github.com/gravitational/teleport/pull/19789.

This is the final PR in the chain. Fixes https://github.com/gravitational/teleport/issues/17257.

---

**TODO:** Testing: 
- [x] Google Storage interaction with `gsutil` [tool](https://cloud.google.com/storage/docs/gsutil). PR: https://github.com/gravitational/teleport/pull/20072
- [x] Terraform GCP provider. This works out of the box after configuring Terraform.

---

Setup:
1. Create a service account `teleport-service`. 
2. Assign `Service Account User` role to the `teleport-service` account.
3. Deploy Teleport instance to GCP VM. Select `teleport-service` as a VM account to use. If you cannot select the service account, check that it has appropriate role (see step 2).
4. Create service accounts for users to use, for example:`teleport-user-1` and `teleport-user-2`. Assign desired permissions for these accounts. 
5. Grant access for `teleport-user-1`, `teleport-user-2` etc. to `teleport-service` at least with `Service Account Token Creator` role. This will allow `teleport-service` to issue tokens as `teleport-user-XXX`. More info: https://cloud.google.com/iam/docs/service-accounts#token-creator-role 
6. Configure GCP as a Teleport application. `cloud: GCP` is the critical part of the config.
```yaml
app_service:
  enabled: "yes"
  apps:
  - name: gcpapp
    cloud: GCP
```
7. Ensure the user is allowed to use chosen service accounts. If this is a fresh cluster, the `access` role should contain the snippet below. If not, modify a role the user has access to include a reference to `{{internal.gcp_service_accounts}}`. You can also hardcode the list of identities in the role itself (not recommended).
```yaml
spec:
  allow:
    gcp_service_accounts:
    - '{{internal.gcp_service_accounts}}'
```
10. When using `{{internal.gcp_service_accounts}}`, the list of allowed roles is sourced from user-specific traits. To update an existing user use `tctl users update <username> --set-gcp-service-accounts ID_1,ID_2,ID_3,...`, where `ID_N` is a full name of service account (e.g. `teleport-user-1@my-account-123456.iam.gserviceaccount.com` )
11. Refresh client login with `tsh logout; tsh login --proxy=...`.
12. Log in to `gcpapp` app with `tsh app login gcpapp [--gcp-service-account ACCOUNT]`. The service account is optional if there is just one available. The service account name can be shortened by discarding domain (e.g. `teleport-user-1`) as long as it doesn't make it ambigous.
13. Run commands with `tsh gcloud ...`, e.g. `tsh gcloud compute instances list`. 
14. Alternatively, start a proxy with `tsh proxy gcloud` and follow printed instructions.

Notes:
1. Instead of deploying Teleport to GCP VM, you can download the key for `teleport-service` service account to your local machine and set the `GOOGLE_APPLICATION_CREDENTIALS` env variable to its location. This is handful in development, but not recommended for production as the key file can be stolen. Example: `GOOGLE_APPLICATION_CREDENTIALS=/path/to/teleport-service-my-account-123456-ee0011223344.json teleport start`
2. The setup above assumes Teleport is running as a privileged account `teleport-service` which is inaccessible to Teleport users.

Tener committed:
GCP CLI support: CA changes

Tener committed:
GCP CLI support: auth changes

Tener committed:
GCP CLI support: `tsh` and `tctl` changes

Tener committed:
GCP CLI support: API changes

Tener committed:
GCP CLI support: RBAC changes

Tener committed:
GCP CLI support: server-side request handler

Tener committed:
Review: use specialized functions in tests.

Tener committed:
Review: perform case-insensitive match, add tests, add docs.

Tener committed:
Review: improve docs.

Tener committed:
Review: simplify test.

Tener committed:
Bugfix: pass GCPServiceAccount in CreateAppSessions.

GavinFrazar commented on lib/srv/alpnproxy/gcp_middleware.go:
```diff
@@ -0,0 +1,61 @@
+
+	expectedAuth := "Bearer " + m.Secret
+
+	if auth != expectedAuth {
```
This check is in the local proxy, so it probably doesn't *really* matter, but we should do constant time comparison:
```
subtle.ConstantTimeCompare([]byte(auth), []byte(expectedAuth))
```

Tener replied:
Huh. I wonder how practical such an attack is, but this is good practice indeed.

GavinFrazar commented on tool/tsh/gcp.go:
```diff
@@ -0,0 +1,443 @@
+		return secret, nil
+	}
+
+	return utils.CryptoRandomHex(10)
```
Why 10? Elsewhere we use CryptoRandomHex I see we use at least 16 and up to 32, but idk if it actually matters here. We have `TokenLenBytes` defined in lib/auth/auth.go as 16 as well.

Tener replied:
`auth.TokenLenBytes` is a good default, I'll use it instead.

GavinFrazar commented on tool/tsh/gcp.go:
```diff
@@ -0,0 +1,443 @@
+
+// getGCPSecret will return fresh secret to use or read it from environment.
+func getGCPSecret() (string, error) {
+	secret := os.Getenv("TELEPORT_GCP_SECRET")
```
I think we should make this a constant in constants.go with other env vars we define

Tener replied:
`tsh` has a bunch of env vars defined for its own use only, it is rightfully in `tsh.go`. I'll add an entry there.

GavinFrazar commented on tool/tsh/gcp.go:
```diff
@@ -0,0 +1,443 @@
+func (a *gcpApp) GetEnvVars() (map[string]string, error) {
+	envVars := map[string]string{
+		// TODO: verify secret in calls somehow
+		"CLOUDSDK_AUTH_ACCESS_TOKEN": a.secret,
+
+		// Needed for gcloud CLI to accept our certs.
+		// This isn't portable and applications may have to set different env variables,
+		// add the application cert to system root store (not recommended, ultimate fallback)
+		// or use equivalent of --insecure flag.
+		"REQUESTS_CA_BUNDLE": a.profile.AppLocalCAPath(a.app.Name),
```
nit: add reference URL link to documentation for these env vars if possible

Tener replied:
Trying to find the docs for `REQUESTS_CA_BUNDLE` I realized it is non-standard env var and isn't documented anywhere. Instead, there is [`core.custom_ca_certs_file`](https://cloud.google.com/sdk/gcloud/reference/config/set#:~:text=custom_ca_certs_file) property which can be set via env variables too (`CLOUDSDK_CORE_CUSTOM_CA_CERTS_FILE`).

GavinFrazar commented on tool/tsh/gcp.go:
```diff
@@ -0,0 +1,443 @@
+	return len(s)
+}
+
+// Less compares items by name.
+func (s SortedGCPServiceAccounts) Less(i, j int) bool {
+	beforeI, afterI, _ := strings.Cut(s[i], "@")
+	beforeJ, afterJ, _ := strings.Cut(s[j], "@")
+
+	if afterI != afterJ {
+		return afterI < afterJ
+	}
+
+	return beforeI < beforeJ
+}
```
Could you add docs describing the expected behavior of this sorting? I think it's sorting first by what follows @ and then by what precedes "@" if the suffix is equal?

Tener replied:
Good point, will do.

GavinFrazar commented on tool/tsh/gcp.go:
```diff
@@ -0,0 +1,443 @@
+	}
+	gcpApps := getGCPApps(profile.Apps)
+	if len(gcpApps) == 0 {
+		return nil, trace.NotFound("Please login to GCP App using 'tsh app login' first")
```
nit: 
```suggestion
		return nil, trace.NotFound("Please login to a GCP App using 'tsh app login' first")
```

GavinFrazar reviewed (changes requested):
Mostly LGTM, but a few things I'd like to address before approving

Tener commented on tool/tsh/app.go:
```diff
@@ -200,6 +219,14 @@ Your identity: {{.identity}}
+var gcpCliTpl = template.Must(template.New("").Parse(
+	`Logged into GCP app "{{.appName}}".
+Your service account: {{.serviceAccount}}
+Example command: tsh gcloud compute instances list
```
`gcloud` is the nominal name of the CLI binary. Without `tsh` the command `gcloud compute instances list` would work, whereas `gcp compute instances list` would not. I didn't realize it doesn't show up in `tsh -h`. I'll make `gcloud` the main command and `gcp` the alias.

Tener commented on tool/tsh/gcp.go:
```diff
@@ -0,0 +1,443 @@
+		return trace.Wrap(err)
+	}
+
+	// // backend expects the tokens to be signed with web session private key
+	// ws, err := tc.GetAppSession(a.cf.Context, types.GetAppSessionRequest{SessionID: a.app.SessionID})
+	// if err != nil {
+	// 	return trace.Wrap(err)
+	// }
+
+	// wsPK, err := utils.ParsePrivateKey(ws.GetPriv())
+	// if err != nil {
+	// 	return trace.Wrap(err)
+	// }
```
Ah, indeed. Thanks.

Tener committed:
Review: use `ConstantTimeCompare` when checking secret value.

Tener committed:
Review: use plural form `GCPServiceAccountsASN1ExtensionOID`

Tener committed:
Review: improve GCP handler.

- make debug log entry in appropriate place
- document the default scope list
- replace `context.WithTimeout` with `s.Clock.After` to avoid `time.Sleep` in tests.
- cloud client is now being mocked, making `generateAccessToken` testeable.
- update tests

Tener committed:
Review: improve Azure handler.

- replace `context.WithTimeout` with `s.Clock.After` to avoid `time.Sleep` in tests.

Tener committed:
Review: improve `tsh gcloud` and `tsh proxy gcloud`

- make `tsh gcloud` and `tsh proxy gcloud` base commands so they are shown in `--help`
- `tsh gcp` and `tsh proxy gcp` are aliases now
- switch `REQUESTS_CA_BUNDLE` to official `CLOUDSDK_CORE_CUSTOM_CA_CERTS_FILE`
- add doc links for `CLOUDSDK_AUTH_ACCESS_TOKEN` and `CLOUDSDK_CORE_CUSTOM_CA_CERTS_FILE`
- remove commented out code
- add docs and tests for `SortedGCPServiceAccounts`
- move `TELEPORT_GCLOUD_SECRET` to constant
- typos

smallinsky commented on lib/srv/alpnproxy/gcp_middleware.go:
```diff
@@ -0,0 +1,62 @@
+	"github.com/sirupsen/logrus"
+)
+
+// GCPMiddleware implements a simplified version of MSI server serving auth tokens.
```
I think that **MSI** is Azure specific name and doesn't apply to Google Cloud

Tener replied:
Fixed docs here and elsewhere.

Tener commented on lib/srv/alpnproxy/gcp_middleware.go:
```diff
@@ -0,0 +1,62 @@
+)
+
+// GCPMiddleware implements a simplified version of MSI server serving auth tokens.
+type GCPMiddleware struct {
```
Yeah, I guess this isn't really GCP specific anymore.

Tener replied:
Renamed, added comments.

Tener commented on lib/srv/alpnproxy/gcp_middleware.go:
```diff
@@ -0,0 +1,62 @@
+	return nil
+}
+
+func (m *GCPMiddleware) HandleRequest(rw http.ResponseWriter, req *http.Request) bool {
```
Middleware can do much more than just return an error. For Azure this is where MSI is implemented: https://github.com/gravitational/teleport/blob/3ebbfdd222eb5111aebe17d079d694d6c08776ec/lib/srv/alpnproxy/azure_msi_middleware.go#L81-L133

`HandleRequest` can be seen as an event handler. Returning a `bool` from event handler to indicate if the event was handled or not is a pretty standard convention.

Tener replied:
Docs added.

Tener committed:
Review: rename GCPMiddleware, fix docs, return precise error.

Tener commented on lib/srv/alpnproxy/gcp_middleware.go:
```diff
@@ -0,0 +1,62 @@
+
+var _ LocalProxyHTTPMiddleware = &GCPMiddleware{}
+
+func (m *GCPMiddleware) CheckAndSetDefaults() error {
```
Added.

greedy52 commented on tool/tsh/gcp.go:
```diff
@@ -0,0 +1,439 @@
+// The request flow to remote server (i.e. GCP APIs) looks like this:
+// clients -> local forward proxy -> local ALPN proxy -> remote server
+//
+// However, with MSI_ENDPOINT variable set, clients will reach out to this address for tokens.
```
nit: update doc

greedy52 reviewed (approved):
LGTM. And great setup guide 👍

Tener committed:
Review: update stale doc.

Tener committed:
Merge branch 'master' into tener/gcloud-cli-tsh-tctl

Tener commented:
@smallinsky 

Regarding point:
> - [x] Terraform GCP provider.

This works out of the box, but Terraform needs to be configured correctly.

Assuming we are running `tsh proxy gcp` like so:

```
TELEPORT_GCLOUD_SECRET=my_gcloud_secret tsh proxy gcp --port 3500
Started GCP proxy on http://127.0.0.1:3500.

Use the following credentials and HTTPS proxy setting to connect to the proxy:

  export BOTO_CONFIG=/Users/myuser/.tsh/gcp/teleport.example.com/gcp/bd460f17_boto.cfg
  export CLOUDSDK_AUTH_ACCESS_TOKEN=my_gcloud_secret
  export CLOUDSDK_CONFIG=/Users/myuser/.tsh/gcp/teleport.example.com/gcp/gcloud
  export CLOUDSDK_CORE_CUSTOM_CA_CERTS_FILE=/Users/myuser/.tsh/keys/teleport.example.com/myuser-app/teleport.example.com/gcp-localca.pem
  export CLOUDSDK_CORE_PROJECT=myproject-123456
  export HTTPS_PROXY=http://127.0.0.1:3500
```

Then we must run Terraform with the following env variables exported:

```
  export HTTPS_PROXY=http://127.0.0.1:3500
  export GOOGLE_OAUTH_ACCESS_TOKEN=my_gcloud_secret
```

Instead of using `GOOGLE_OAUTH_ACCESS_TOKEN` the token can be saved in a file as per https://registry.terraform.io/providers/hashicorp/google/latest/docs/guides/provider_reference#access_token. 

Unlike some other apps, Terraform does not provide explicit support for custom CA certs: https://github.com/hashicorp/terraform/issues/28551. 

To make it trust `/Users/myuser/.tsh/keys/teleport.example.com/myuser-app/teleport.example.com/gcp-localca.pem` we must use platform-specific mechanisms available to Go-based programs. On Linux that can be `SSL_CERT_FILE` env var, on MacOS `security add-trusted-cert`. I'm not sure what the preferred mechanism is on Windows.
